proposals/004-skips/README.md

```diff
@@ -0,0 +1,179 @@
+
+Owner: jiazh@microsoft.com
+
+Impacted Projects:
+
+- [x] spin-operator
+- [ ] `spin kube` plugin
+- [ ] runtime-class-manager
+- [ ] containerd-shim-spin
+- [ ] Governance
+- [ ] Creates a new project
+
+Created: October 16th, 2024
+
+Updated: October 16th, 2024
+
+## Background
+
+The [Runwasi] project supports running Wasm workloads in the same pod as Linux containers. Currently, the SpinApp Custom Resource Definition (CRD) does not support running [Sidecar Containers]. Sidecars are important in use cases where additional services need to run in the same pod as the Spin application, enabling low-latency communication and resource sharing.
+
+[Sidecar Containers]: (https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/)
+[Runwasi]: (https://github.com/containerd/runwasi)
+[SpinApp CRD]: (https://www.spinkube.dev/docs/reference/spin-app/)
+[Spin Operators repository]: (https://github.com/spinkube/spin-operators)
+
+### User Stories
+
+- As an user, I have a proprietary solution packaged as a binary that cannot be ported to Wasm. I would like to run this binary as a Linux container. Also, I would like to run this container in the same pod as the Spin application so that they can communicate with each other locally. In addition, I want the sidecar container to start before the Spin application.
+- As a developer, I would like to run [Dapr] sidecar containers in the same pod as the Spin application so that I can leverage Dapr's capabilities, such as service invocation, state management, pub/sub, etc, in my Spin application.
+- As a developer, I would like to run a [Envoy] proxy sidecar container in the same pod as the Spin application so that I can re-use the service mesh capabilities provided by Envoy, such as traffic management, security, observability, etc, in my Spin application.
+
+[Dapr]: (https://dapr.io/)
+[Envoy]: (https://www.envoyproxy.io/)
+
+### Existing applications
+
+Here are some existing prototypes that I have found that runs Wasm workloads side by side with Linux containers:
+
+- [SpinKube - the first look at WebAssembly/WASI application (SpinApp) on Kubernetes](https://dev.to/thangchung/spinkube-the-first-look-at-webassemblywasi-application-spinapp-on-kubernetes-36jd) and [Coffeeshop repo](https://github.com/thangchung/coffeeshop-on-spinkube/tree/feat/global-azure-2024)
+- [Istio Wasm Demo](https://github.com/keithmattix/istio-wasm-demo)
+
+
+## Proposal
+
+
+The main changes are
+
+1. A new field `InitContainers` in the `spinapp_types.go` that defines a list of sidecar containers to be included in the Deployment.
+```go
+// InitContainers defines the list of sidecar containers to be included in the deployment.
+//
+// These containers will not include the main Spin App. They share the Spin App's
+// environment variables and volumes.
+// +kubebuilder:validation:Optional
+InitContainers []corev1.Container `json:"containers,omitempty"`
```

Comment:
Doesn't the term init container imply that the container is short lived and for setup — not long lived and running as a sidecar? https://kubernetes.io/docs/concepts/workloads/pods/init-containers/

Reply:
> Kubernetes implements sidecar containers as a special case of [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/); sidecar containers remain running after Pod startup. This document uses the term regular init containers to clearly refer to containers that only run during Pod startup.

Saw this later

Reply:
^^ 
Although there is reasoning why `InitContainers` would be the correct term, I would suggest using two dedicated terms instead: e.g. `Initializers` and `Sidecars`. 

I would also prefer embedding the `corev1.Container` struct into a higher level struct, allowing users to specify the desired runtime class. This would unlock the capability to run either:

- Spin Apps (e.g. using the command trigger) or traditional containers during initialization 
- Spin Apps or traditional containers as sidecars.

```yaml
apiVersion: core.spinoperator.dev/v1alpha1
kind: SpinApp
metadata:
  name: simple-spinapp
spec:
  image: "ttl.sh/spinapp:44h"
  replicas: 1
  executor: containerd-shim-spin
  volumes:
    - name: example-volume
      persistentVolumeClaim:
        claimName: example-pv-claim
  volumeMounts:
    - name: example-volume
      mountPath: "/mnt/data"
  initializers:
    - name: my-init-container
      runtimeClass: default # (traditional container) could also be omitted
      spec:
        image: busybox:1.28
        command: ['sh', '-c', "until nslookup mydb.$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace).svc.cluster.local; do echo waiting for mydb; sleep 2; done"]
  sidecars:
    - name: spin-sidecar
      runtimeClass: wasmtime-spin-v2
      spec:
        image: "ttl.sh/spin-logshipper:44h"
```

Reply:
I like the idea of splitting init containers and sidecar containers.

I feel a bit weird about directly specifiying the `runtimeClass` on the init/sidecar containers because we don't do that for the root SpinApp, rather we refer to an `executor` which references the `runtimeClass`. This would logically lead us to specifying `executor` for the init/sidecar containers which quickly falls apart because we need all these to run in the same pod.

I guess as I think about it I think we should just force init/sidecar containers to use the same runtime class as the root spin app. I believe the shim is smart enough to just do that.

Reply:
I am not sure if I understand the motivation to specify `runtimeClass` for Sidecars and initializers here. AFAIU, RuntimeClass should be specified to a Pod, instead of individual containers. Also, as @calebschoepp described, the SpinApp has a `executor` that maps to a RuntimeClass in the Pod. 

To splitting init containers and sidecar containers, I am concerned that it will increase the learning curve for users if they are already familiar with the sidecar container concept in Kubernetes, which use the `initContainers ` syntax.

Reply:
+1 to using `initContainers` terminology. @Mossaka is there any case where a k8s init container could not be run on the shim? I am guessing no. Want to make sure we can say "anything that is an init container (and doesnt use port 80) cna go here"

Reply:
The expectation for the shim is that it run can run any containers that comform to the OCI Spec, including the init containers. If in pactice we find there are init containers that cannot be run in the Spin shim, we should regard it as a bug.